tools/etcd-dump-logs/main.go

```diff
@@ -166,3 +130,139 @@ func excerpt(str string, pre, suf int) string {
+	var r etcdserverpb.Request
+	if err := r.Unmarshal(entry.Data); err == nil {
+		fmt.Printf("%4d\t%10d\tnorm", entry.Term, entry.Index)
+		switch r.Method {
+		case "":
+			fmt.Printf("\tnoop\n")
+		case "SYNC":
+			fmt.Printf("\tmethod=SYNC time=%q\n", time.Unix(0, r.Time))
+		case "QGET", "DELETE":
+			fmt.Printf("\tmethod=%s path=%s\n", r.Method, excerpt(r.Path, 64, 64))
+		default:
+			fmt.Printf("\tmethod=%s path=%s val=%s\n", r.Method, excerpt(r.Path, 64, 64), excerpt(r.Val, 128, 0))
+		}
+	}
+}
+
+// printInternalRaftRequest is used to print entry information for InternalRaftRequestRange, InternalRaftRequestPut,
+// InternalRaftRequestDeleteRange and InternalRaftRequestTxn entries
+func printInternalRaftRequest(entry raftpb.Entry) {
+	var rr etcdserverpb.InternalRaftRequest
+	if err := rr.Unmarshal(entry.Data); err == nil {
+		fmt.Printf("%4d\t%10d\tnorm\t%s\n", entry.Term, entry.Index, rr.String())
+	}
+}
+
+// evaluateEntrytypeFlag evaluates entry-type flag and choose proper filter/filters to filter entries
+func evaluateEntrytypeFlag(entrytype string) []EntryFilter {
+	var entrytypelist []string
+	if entrytype != "" {
+		entrytypelist = strings.Split(entrytype, ",")
+	}
+
+	validRequest := map[string][]EntryFilter{"ConfigChange": {passConfChange},
+		"Normal":                         {passInternalRaftRequestRange, passInternalRaftRequestPut, passInternalRaftRequestDeleteRange, passInternalRaftRequestTxn, passRequest},
+		"Request":                        {passRequest},
+		"InternalRaftRequest":            {passInternalRaftRequestRange, passInternalRaftRequestPut, passInternalRaftRequestDeleteRange, passInternalRaftRequestTxn},
+		"InternalRaftRequestRange":       {passInternalRaftRequestRange},
+		"InternalRaftRequestPut":         {passInternalRaftRequestPut},
+		"InternalRaftRequestDeleteRange": {passInternalRaftRequestDeleteRange},
+		"InternalRaftRequestTxn":         {passInternalRaftRequestTxn}}
+	filters := make([]EntryFilter, 0)
+	if len(entrytypelist) == 0 {
+		filters = append(filters, passConfChange)
+		filters = append(filters, passInternalRaftRequestRange)
+		filters = append(filters, passInternalRaftRequestPut)
+		filters = append(filters, passInternalRaftRequestDeleteRange)
+		filters = append(filters, passInternalRaftRequestTxn)
+		filters = append(filters, passRequest)
+	}
+	for _, et := range entrytypelist {
+		if f, ok := validRequest[et]; ok {
+			filters = append(filters, f...)
+		} else {
+			log.Printf(`[%+v] is not a valid entry-type, ignored. 
+Please set entry-type to one or more of the following: 
+ConfigChange, Normal, Request, InternalRaftRequest, 
+InternalRaftRequestRange, InternalRaftRequestPut, InternalRaftRequestDeleteRange, InternalRaftRequestTxn`, et)
+		}
+	}
+
+	return filters
+}
+
+//  listEntriesType filters and prints entries based on the entry-type flag,
+func listEntriesType(entrytype string, ents []raftpb.Entry) {
+	entryFilters := evaluateEntrytypeFlag(entrytype)
```

Comment:
Can `evaluateEntrytypeFlag` handle `entrytype == ""`?

I double-checked out code, and lease grant does get recorded in WAL. And `etcd-dump-logs ./data-dir` with no `--entry-type` flag would omit those entries, while current master branch `etcd-dump-logs` tool prints them out. If `--entry-type` is empty, it should prints out all entries by default (e.g. lease grant, revoke, etc.).

Thanks!

Reply:
Yes, you are right! Thanks for the catch! I have fixed that and commented in place.